### ISSUE TITLE
[stable-1.2] Support bundling with Qt4

### ISSUE THREAD

**Issue by barracuda156** [LMMS/lmms#6793]
@tresf This is a practical issue, our Macports fixes for bundling script were done against an archaic version, and substantial changes occurred since then in the sources.
Currently bundling script assumes Qt5: https://github.com/LMMS/lmms/blob/stable-1.2/cmake/apple/install_apple.sh.in
Since `stable-1.2` supports Qt4 otherwise, I hope this can be fixed :)

In particular, I am not sure what to do about `Carla`: it is supposed to be a standalone (?) app, but there is no code there to create an app bundle for it.

I also do not get why this is done: https://github.com/LMMS/lmms/blob/d87917cdc9d4777784a643163cbd5f264079c9b6/cmake/apple/install_apple.sh.in#L31
As is, it tries to write directly to prefix, which obviously fails due to permissions, but even if changed to destroot, it makes no sense to dump a random plist into the root directory. Perhaps it should go instead into the app bundle, together with other stuff?

P. S. In the worst case we can implement bundling from scratch on our end, provided I have a ready package (so that I can see what goes where), but that will leave the problem unsolved for anyone trying to use 1.2 without Macports, but with Qt4.

**Comment by tresf** [comment]
> @tresf This is a practical issue, our Macports fixes for bundling script were done against an archaic version, and substantial changes occurred since then in the sources. Currently bundling script assumes Qt5: https://github.com/LMMS/lmms/blob/stable-1.2/cmake/apple/install_apple.sh.in Since `stable-1.2` supports Qt4 otherwise, I hope this can be fixed :)

It only assumes `brew` and `qt5` if it **cannot find `macdeployqt`** per: 

https://github.com/LMMS/lmms/blob/d87917cdc9d4777784a643163cbd5f264079c9b6/cmake/apple/install_apple.sh.in#L21-L25C3

You may -- if you wish -- use an `alias` to set `macdeployqt` and then it will skip that assumption.  If you'd like to detect Qt4 automatically I will need some help.  For starters, can you try:

```bash
echo "################# @QT_QMAKE_EXECUTABLE@"
```

This may help us find the path of `macdeployqt`, e.g.

```diff
 # Locate macdeployqt, assume homebrew & Qt5
 which macdeployqt > /dev/null 2>&1
  if [ $? -ne 0 ]; then
    export PATH=$PATH:$(brew --prefix qt)/bin
+ else
+  # Calculate from qmake path
+  export PATH=$PATH:$(dirname "@QT_QMAKE_EXECUTABLE@")
+ fi
```

One could argue that this approach should be preferred for ALL Qt versions since it doesn't rely on a pre-set `PATH`.

> In particular, I am not sure what to do about `Carla`: it is supposed to be a standalone (?) app, but there is no code there to create an app bundle for it.

Yes, it's an optional, standalone app.  If that line throws errors, we'll need to work from the error message.  If it does not, I would ignore this.
 
> I also do not get why this is done:
> ```bash
>  \cp  "@CMAKE_BINARY_DIR@/Info.plist" "@CMAKE_INSTALL_PREFIX@/" 
> ```
> As is, it tries to write directly to prefix, which obviously fails due to permissions, but even if changed to destroot, 

Sorry, the installer will only work if `CMAKE_INSTALL_PREFIX` is set.  We usually use `../target` as a value, e.g. from `build`: `cmake .. -DCMAKE_INSTALL_PREFIX=../target`.

> it makes no sense to dump a random plist into the root directory. Perhaps it should go instead into the app bundle, together with other stuff?

The root directory becomes the app bundle.

> P. S. In the worst case we can implement bundling from scratch on our end, provided I have a ready package (so that I can see what goes where), but that will leave the problem unsolved for anyone trying to use 1.2 without Macports, but with Qt4.

Please let me know how the above feedback helps.  I think we can support your configuration with minimal effort.

**Comment by barracuda156** [comment]
> It only assumes brew and qt5 if it cannot find macdeployqt

There were [issues](https://trac.macports.org/ticket/36431) with `macdeployqt` in Qt4, and Macports [removed](https://github.com/macports/macports-ports/blob/c3a496cd95af77e4a880e77767541b4d7afcc172/multimedia/lmms/files/patch-data-scripts-create_apple_bundle.sh.in.diff#L42-L44) dependency on it earlier, but I will try. Maybe it got fixed eventually.

Or likely it will work without it too, just adding a fix below.

> Sorry, the installer will only work if `CMAKE_INSTALL_PREFIX` is set.

Macports sets it (which can be disabled if needed). But usually `CMAKE_INSTALL_PREFIX` is _the prefix,_ so at destroot it stands for `${destroot}${prefix}` (default value is `${destroot}/opt/local`), and then everything gets relocated to appropriate locations inside the prefix. But after your comments I can relate to what the earlier Macports patch implied: https://github.com/macports/macports-ports/blob/c3a496cd95af77e4a880e77767541b4d7afcc172/multimedia/lmms/Portfile#L61-L69
Looks like that hackery is still relevant.

> Please let me know how the above feedback helps. I think we can support your configuration with minimal effort.

Greatly appreciated!

**Comment by tresf** [comment]
> Macports sets it (which can be disabled if needed). But usually CMAKE_INSTALL_PREFIX is the prefix, so at destroot it stands for ${destroot}${prefix} (default value is ${destroot}/opt/local), and then everything gets relocated to appropriate locations inside the prefix. But after your comments I can relate to what the earlier Macports patch implied: [...] Looks like that hackery is still relevant.

Agreed and in hindsight, I realize these two **should NOT be coupled together**,  but to decouple these steps you would probably want to remove the `install` script entirely.

So I need to know... What is your end-goal?  Do you want a usable, portable `.app` or do you simply want `lmms` to startup from command line?  If the former, I would try to keep the workarounds that Macports currently uses.  If the latter, I would recommend ditching the installer script, or perhaps moving it to be a "bundler" script and cherry-pick the relevant parts.  I'm fine with either, but it's been a while since I've done custom steps with cmake, so it'll be a relearning exercise for me.  To decouple:
* All CI jobs need to be aware of this, additional cmake targets must be added to CI jobs
* Build wikis must be updated
* Both Macports as well as other package managers like Homebrew should be put on alert so the recipes can be updated
* Patches should land on `master` first, then backported

> > It only assumes brew and qt5 if it cannot find macdeployqt
> 
> There were [issues](https://trac.macports.org/ticket/36431) with `macdeployqt` in Qt4, and Macports [removed](https://github.com/macports/macports-ports/blob/c3a496cd95af77e4a880e77767541b4d7afcc172/multimedia/lmms/files/patch-data-scripts-create_apple_bundle.sh.in.diff#L42-L44) dependency on it earlier, but I will try. Maybe it got fixed eventually.

The comments on that patch seem like it's fixed.  Have you tried?  Furthemore, this effort will largely depend on whether or not you want a distributable, installable app.  I may be easiest to just leverage `macdeployqt` and fix any bugs it has manually.

**Comment by barracuda156** [comment]
@tresf Sure, since app bundle is implemented, that is preferable. It is more user-friendly and looks better.
I will see if `macdeployqt ` is usable in Qt4.

**Comment by barracuda156** [comment]
@tresf Here:
![lmms_installed](https://github.com/LMMS/lmms/assets/92015510/872fd042-819e-461c-80f9-a89f589c638b)

**Comment by barracuda156** [comment]
UPD. @tresf Okay, looks like https://github.com/LMMS/lmms/pull/6794 fixes arch display in About LMMS:
![lmms_cpu_fixed](https://github.com/LMMS/lmms/assets/92015510/1d56ad2f-dc24-42f9-845c-e19f787fbe01)

**Comment by tresf** [comment]
> @tresf Good news: old Macports fix for destroot worked perfectly. I have the app now installed.
> 
> Two minor issues:
> 
> 1. About LMMS still depicts `OS X/unknown processor`, despite configure recognizing correctly `powerpc`. Where should it take the correct value from?

This questions seems to be answered in #6794.

> UPD. Never mind point 1, I have found it – it is in `versioninfo.h`. Commit that we have backported missed that part, so it was not fixed. Surprisingly, it is not there even in the master:

Please, if you can, port this patch forward into `master` so that new versions also take advantage of this fix.


> 2. App icon does not display, though this could be 10.6 issue. I cannot check on 10.5, no environment.

There's a chance that the `icon.icns` file is incompatible with this system, however in my experience this is very rare.  Instead, I would recommend the following:

* Reboot.  Mac AGGRESSIVELY caches the app information, rebooting fixes these things quite often.
* If rebooting doesn't fix it, try to replace the `icon.icns` file with a known good file (such as the one from `TenFourFox.app`, or whatever).
* If a new `icns` file doesn't fix it, play around with `Info.plist`, specifically this value:
   ```xml
       <key>CFBundleIconFile</key>
       <string>icon.icns</string>
   ```
   Double-check this plist against other known working plist files and see if there's a mistake we made.

**Comment by barracuda156** [comment]
I have a couple of apps with broken icons (just emptiness) – `R GUI` and just-today-fixed-for-ppc `avidemux2`. Most of apps are fine. Another two only got generic Apple app icon (i.e. visible, but not the app original one). I suspect that either `icns` is unsupported on older systems, or its HD version, of possibly there is a rendering bug in 10.6 PPC.

In any case this is not something of a real concern and likely can be indeed fixed by simple swapping. Unless confirmed that the same issue affects 10.5.8, no need to do anything about it. If it does, then maybe a fallback icon can be made.
